### PR TITLE
Fix potential out of bounds access in Production/ResearchQueue

### PR DESCRIPTION
**Pull request by Morlic-fo**
It was possible to access the ProductionQueue/ResearchQueue out of bounds in Release mode. For example, following test wouldn't necessarily raise an error (but possibly crash the AI) before the fix. Now, as expected, an IndexError is raised instead.
```
[03 Mai 22:29:36] AI_1: Entering debug mode
Print 'stop' to exit.
Local variables:
  ai  aistate
  u   universe
  e   empire
[03 Mai 22:29:45] Menschlicher_Spieler: e.productionQueue[999]
[03 Mai 22:29:45] AI_1: Traceback (most recent call last):
  File "<input>", line 1, in <module>
IndexError: Tried to access ProductionQueue element out of bounds
```

### CODE REVIEW

**Comment by o01eg** [comment]
Does it fix  #2412? I suppose issue with using index to address item in the queues while it depends on orders' sorting.

File: Empire/ResearchQueue.cpp [hunk 1]
```diff
@@ -323,7 +324,8 @@ void ResearchQueue::insert(iterator it, const std::string& tech_name, bool pause
 { m_queue.insert(it, Element(tech_name, m_empire_id, 0.0f, -1, paused)); }
 
 void ResearchQueue::erase(iterator it) {
-    assert(it != end());
+    if (it == end())
```
Comment (geoffthemedio) [inline Empire/ResearchQueue.cpp:327]: is this necessarily a crash and/or preferable to throw for, rather than it being a no-op?

Comment (Morlic-fo) [reply Empire/ResearchQueue.cpp:327]: I have no opinion on this. I implemented the throw considering the previous assert statement. If you think a no-op return would make more sense here, I could change it.

**Comment by Morlic-fo** [comment]
> Does it fix #2412?

I don't think so. It may or may not throw an std::out_of_range exception with this PR rather than an assertion failure or the seemingly optimized out [std::runtime_error](https://github.com/freeorion/freeorion/blob/master/Empire/Empire.cpp#L1148).

Without exception handling, I would still expect a crash.

**Pull request merged**